From 79c7c01ce2486032202d72c9c8120836adf0dafa Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Mon, 4 Sep 2023 16:34:22 +0200
Subject: [PATCH 0399/1695] drivers: stm32_rng: implement clock support for
 STM32MP25x platforms

Implement the clock support for STM32MP25x platforms. On this platform,
a security clock is shared between some hardware blocks. For the RNG,
it is the RNG kernel clock. Therefore, the gate is no more shared
between the RNG bus and kernel clocks as on STM32MP1x platforms and
the bus clock has to be managed on its own.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git e913d627024505d3c7da8314e98a74fb9b084649 drivers: stm32_rng: implement clock support for STM32MP25x platforms]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: Iff65b1ad8548ecefef1ae85c5530b100eb6e72c1
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/325750
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/stm32_rng.c | 55 +++++++++++++++++++++++++++++++++++++---
 1 file changed, 52 insertions(+), 3 deletions(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index f6111e08f..5c812379d 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -71,6 +71,7 @@
 
 struct stm32_rng_driver_data {
 	unsigned long max_noise_clk_freq;
+	unsigned long nb_clock;
 	bool has_power_optim;
 	bool has_cond_reset;
 	bool entropy_src_config;
@@ -79,6 +80,7 @@ struct stm32_rng_driver_data {
 struct stm32_rng_instance {
 	struct io_pa_va base;
 	struct clk *clock;
+	struct clk *bus_clock;
 	struct rstctrl *rstctrl;
 	const struct stm32_rng_driver_data *ddata;
 	unsigned int lock;
@@ -374,6 +376,14 @@ static TEE_Result stm32_rng_read(uint8_t *out, size_t size)
 	if (rc)
 		return rc;
 
+	if (stm32_rng->bus_clock) {
+		rc = clk_enable(stm32_rng->bus_clock);
+		if (rc) {
+			clk_disable(stm32_rng->clock);
+			return rc;
+		}
+	}
+
 	rng_base = get_base();
 
 	/* Arm timeout */
@@ -412,6 +422,8 @@ static TEE_Result stm32_rng_read(uint8_t *out, size_t size)
 out:
 	assert(!rc || rc == TEE_ERROR_GENERIC);
 	clk_disable(stm32_rng->clock);
+	if (stm32_rng->bus_clock)
+		clk_disable(stm32_rng->bus_clock);
 
 	return rc;
 }
@@ -525,12 +537,22 @@ stm32_rng_pm(enum pm_op op, unsigned int pm_hint __unused,
 	if (res)
 		return res;
 
+	if (stm32_rng->bus_clock) {
+		res = clk_enable(stm32_rng->bus_clock);
+		if (res) {
+			clk_disable(stm32_rng->clock);
+			return res;
+		}
+	}
+
 	if (op == PM_OP_RESUME)
 		res = stm32_rng_pm_resume();
 	else
 		res = stm32_rng_pm_suspend();
 
 	clk_disable(stm32_rng->clock);
+	if (stm32_rng->bus_clock)
+		clk_disable(stm32_rng->bus_clock);
 
 	return res;
 }
@@ -556,9 +578,21 @@ static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 	if (res != TEE_SUCCESS && res != TEE_ERROR_ITEM_NOT_FOUND)
 		return res;
 
-	res = clk_dt_get_by_index(fdt, node, 0, &stm32_rng->clock);
-	if (res)
-		return res;
+	if (stm32_rng->ddata->nb_clock > 1) {
+		res = clk_dt_get_by_name(fdt, node, "rng_clk",
+					 &stm32_rng->clock);
+		if (res)
+			return res;
+
+		res = clk_dt_get_by_name(fdt, node, "rng_hclk",
+					 &stm32_rng->bus_clock);
+		if (res)
+			return res;
+	} else {
+		res = clk_dt_get_by_index(fdt, node, 0, &stm32_rng->clock);
+		if (res)
+			return res;
+	}
 
 	if (fdt_getprop(fdt, node, "clock-error-detect", NULL))
 		stm32_rng->clock_error = true;
@@ -617,6 +651,14 @@ static TEE_Result stm32_rng_probe(const void *fdt, int offs,
 	if (res)
 		goto err;
 
+	if (stm32_rng->bus_clock) {
+		res = clk_enable(stm32_rng->bus_clock);
+		if (res) {
+			clk_disable(stm32_rng->clock);
+			goto err;
+		}
+	}
+
 	version = io_read32(get_base() + RNG_VERR);
 	DMSG("RNG version Major %u, Minor %u",
 	     (version & RNG_VERR_MAJOR_MASK) >> RNG_VERR_MAJOR_SHIFT,
@@ -639,6 +681,8 @@ static TEE_Result stm32_rng_probe(const void *fdt, int offs,
 		goto err_clk;
 
 	clk_disable(stm32_rng->clock);
+	if (stm32_rng->bus_clock)
+		clk_disable(stm32_rng->bus_clock);
 
 	if (stm32_rng->release_post_boot)
 		stm32mp_register_non_secure_periph_iomem(stm32_rng->base.pa);
@@ -655,6 +699,8 @@ static TEE_Result stm32_rng_probe(const void *fdt, int offs,
 
 err_clk:
 	clk_disable(stm32_rng->clock);
+	if (stm32_rng->bus_clock)
+		clk_disable(stm32_rng->bus_clock);
 err:
 	free(stm32_rng);
 	stm32_rng = NULL;
@@ -665,6 +711,7 @@ err:
 static const struct stm32_rng_driver_data mp13_data[] = {
 	{
 		.max_noise_clk_freq = U(48000000),
+		.nb_clock = 1,
 		.has_cond_reset = true,
 		.entropy_src_config = true,
 		.has_power_optim = true,
@@ -674,6 +721,7 @@ static const struct stm32_rng_driver_data mp13_data[] = {
 static const struct stm32_rng_driver_data mp15_data[] = {
 	{
 		.max_noise_clk_freq = U(48000000),
+		.nb_clock = 1,
 		.has_cond_reset = false,
 		.entropy_src_config = false,
 		.has_power_optim = false,
@@ -684,6 +732,7 @@ DECLARE_KEEP_PAGER(mp15_data);
 static const struct stm32_rng_driver_data mp25_data[] = {
 	{
 		.max_noise_clk_freq = U(48000000),
+		.nb_clock = 2,
 		.has_cond_reset = true,
 		.entropy_src_config = true,
 		.has_power_optim = true,
-- 
2.47.2

